Preprocess interface + implementations

ASalvail opened the pull request:
First draft of `Preprocess` class and a couple implementations.

Any thoughts @MarcCote ?

ASalvail committed:
First draft of `Preprocess` class and a couple implementation.

ASalvail committed:
Changed preprocessing to enable better chaining and make sure copies are made.

ASalvail commented:
This should do the trick!

MarcCote commented:
This seems perfect but... could you add two small unit tests? One for `NormalizeFeature` and `LagExample`. That would _really_ help me sleep at night :P.

ASalvail committed:
Switched to linked datasets

ASalvail committed:
Added tests for the NormalizeFeature and removed the Lag preprocessor as it wasn't working as intended

ASalvail commented:
Addressed @MarcCote comment. I removed the `LagExample` preprocessor because it really wasn't working as intended. (I actually don't know what it was doing...)

MarcCote commented on smartlearner/preprocessors.py:
```diff
@@ -0,0 +1,32 @@
+from .interfaces.preprocess import Preprocess
+
+
+class NormalizeFeature(Preprocess):
```
We should say what type of normalization it is. According to https://en.wikipedia.org/wiki/Normalization_(statistics) it is "Normalizing errors when population parameters are known. Works well for populations that are normally distributed".

ASalvail committed:
Adds a docstring to NormalizeFeatures
